packages/firebase_auth/lib/firebase_auth.dart

```diff
@@ -73,12 +73,31 @@ class FirebaseAuth {
     'plugins.flutter.io/firebase_auth',
   );
 
+  final List<StreamController<FirebaseUser>> _authStateChangedController =
+      <StreamController<FirebaseUser>>[];
+  int _counter = 0;
+
   /// Provides an instance of this class corresponding to the default app.
   ///
   /// TODO(jackson): Support for non-default apps.
   static FirebaseAuth instance = new FirebaseAuth._();
 
-  FirebaseAuth._();
+  FirebaseAuth._() {
+    channel.setMethodCallHandler(_callHandler);
+  }
+
+  /// Receive [FirebaseUser] each time the user signIn or signOut
+  Stream<FirebaseUser> get onAuthStateChanged {
+    final int id = _counter++;
+
+    _authStateChangedController
+        .add(new StreamController<FirebaseUser>.broadcast(onListen: () {
+      channel.invokeMethod('startListeningAuthState', <String, int>{"id": id});
```

Comment:
You should add the controller to the array here, as opposed to adding it synchronously in the main body. That avoids a situation where someone never listens to the stream returned by `onAuthStateChanged` and we end up leaking the controller. See https://github.com/flutter/plugins/blob/master/packages/firebase_database/lib/src/query.dart#L46 for an example of this pattern.